### PR TITLE
Add native OpenX JSON format

### PR DESCRIPTION
**Pull request by dain**
## Description

Add native implementation of the OpenX JSON SerDe (`org.openx.data.jsonserde.JsonSerDe`) with extensions.  The exact behavior of this format is covered in the README.md in `Add native OpenX JSON format` commit.

## Documentation notes
This format implementation contains several extensions over the original PR as noted in the [README](https://github.com/trinodb/trino/blob/a297801863bf76a30f250eeeffedc39618ece191/lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md) file.  This is a different approach than the other native implementations which strive for a bug-for-bug compatibility.  I think we should summarize the major changes into the main documentations:
* Added support  for reading and writing `DECIMAL(P, S)`, `CHAR(N)`, `VARCHAR(N)`, `VARBINARY`.
* Value parse errors produce a null instead of a failure.
* JSON object or array values for numeric or temporal types, throw an exception.
* Values that are outside the allowed type range result in null.
These extensions are not in Hive and Spark, so any table that relies on these extensions may not be readable and/or writable in Hive and Spark.  In this case, the user would simply change the column types in the Hive Metastore to a type supported by Hive and Spark.

## Release notes

```markdown
# Section
* Add native OpenX JSON file format reader and writer. These can be disabled with the `openx_json_native_reader_enabled` and `openx_json_native_writer_enabled` session properties or the `openx-json.native-reader.enabled` and `openx-json.native-writer.enabled` configuration property. ({16073}``)
```

### CODE REVIEW

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
+# OpenX JSON format
+
+This code in this package implements the OpenX JSON SerDe (`org.openx.data.jsonserde.JsonSerDe`)
+with extensions. The behavior is based on the implementations maintained by
+[Rcongiu](https://github.com/rcongiu/Hive-JSON-Serde) and [Starburst](https://github.com/starburstdata/hive-json-serde).
+The Rcongiu implementation is based on very old Hive systems, has many quarks and bugs, is not
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:6]: "and is not"

Comment (electrum) [reply lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:6]: You mean "quirk" not "quark"

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+actively maintained (as of this writing). The Starburst version is much more modern, and actively
+maintained, but has it own quarks. This implementation generally a super set of both of these
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:8]: "quircks"

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+implementations, including support for reading and writing more types, and generally attempts to
+normalize the behavior of type mappers, and line processing. Except for some noted cases below,
+this implementation will read JSON that the other implementations can read, and can read the JSON
+written by the other implementations. Additionally, the JSON written by this implementation can
+be read by the other implementations, assuming they support the data type at all.
+
+# Line Format
+
+Hive SerDe operate on a per line bases, so multi line JSON can not be supported. Each line is
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:17]: bases -> basis

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+processed as follows:
+
+1. If the first non-whitespace character of the line is not `{` or `]`, a row will all fields
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:20]: will -> with

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+   set to null is produced.
+2. The line is parsed using the very lenient [open-json](https://github.com/tdunning/open-json)
+   parser. A parser failure can be ignored by setting the `ignore.malformed.json` table property,
+   and in this case a row will all fields set to null is produced.
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:24]: will -> with

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+Specific type mapping rules are described below, but, general guidelines are:
+
+* Parse failures result in `null`.
+* Values that are outside the allowed type range result in `null`.
+* JSON object or array values for numeric or temporal types, throw an exception.
+* Text types support all JSON values by rendering the value as canonical JSON.
+
+## BOOLEAN
+
+`true` only if quote or unquoted string is equals ignoring case to `true`. All other values,
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:44]: quote -> quoted

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+### Formats
+
+* Any value that can be parsed by `new BigDecimal(String)`
+* Hex `0x1234`. Negative numbers are not supported.
+* Octal `01234`. Negative numbers are not supported.
+
+### Known Differences
+
+* Rcongiu fails on values with decimals or exponents values.
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:63]: Remove "values" at end of sentence

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+* Starburst performs a narrowing cast to expected size.
+* Starburst and Rcongiu do not support octal in map keys or quoted octal values.
+* Starburst and Rcongiu both have a bug that allows negative hex and octal values formatted at `0x-123` and `0-123`.
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:66]: "formatted like X or Y"

File: lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+* **Object**: value is rendered back to JSON
+
+### Known Differences
+
+* Rcongiu fails on unquoted boolean and numbers in field names, but all other unquoted values are allowed.
+* Starburst and fails for unquoted field names.
+* Rcongiu and Starburst will use `Double` in some cases for canonicalization of numbers which results in
+  a slightly different format from `BigDecimal`.
+* Rcongiu and Starburst do not support conversions for `CHAR` or bounded `VARCHAR`.
+* Rcongiu and Starburst write all `CHAR` or bounded `VARCHAR` values as null.
```
Comment (electrum) [inline lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/openxjson/README.md:117]: This seems to be missing a word or something

**Review by electrum — APPROVED** [review]

**Comment by 1ambda** [comment]
For someone who searching the way to use OpenX JSON Serde, 

Just use OPENX_JSON for ‘formst’ the table property. Those session properties introduced in this pr was removed somehow.

**Comment by electrum** [comment]
cc @mosabua we should document this

**Comment by mosabua** [comment]
@electrum @dain @1ambda - https://github.com/trinodb/trino/pull/22252

**Comment by mosabua** [comment]
Linked PR in preceding comment is merged.

**Pull request merged**